gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -0,0 +1,598 @@
+          .addStateModelDefinition(
+              new StateModelDefinition(StateModelConfigGenerator.generateConfigForTaskStateModel()))
+          .autoJoin(true)
+          .build();
+      boolean clusterCreated = helixConnection.createClusterAccessor(clusterId).createCluster(clusterConfig);
+      if (clusterCreated) {
+        LOGGER.info("Created Helix cluster " + clusterId.stringify());
+      } else {
+        LOGGER.info("Helix cluster " + clusterId.stringify() + " already exists");
+      }
+    } finally {
+      helixConnection.disconnect();
+    }
+  }
+
+  /**
+   * Setup and submit the Gobblin Yarn application.
+   *
+   * @throws IOException if there's anything wrong setting up and submitting the Yarn application
+   * @throws YarnException if there's anything wrong setting up and submitting the Yarn application
+   */
+  private ApplicationId setupAndSubmitApplication() throws IOException, YarnException {
+    YarnClientApplication gobblinYarnApp = this.yarnClient.createApplication();
+    ApplicationSubmissionContext appSubmissionContext = gobblinYarnApp.getApplicationSubmissionContext();
+    ApplicationId applicationId = appSubmissionContext.getApplicationId();
+
+    GetNewApplicationResponse newApplicationResponse =gobblinYarnApp.getNewApplicationResponse();
+    // Set up resource type requirements for ApplicationMaster
+    Resource resource = prepareContainerResource(newApplicationResponse);
+
+    Map<String, LocalResource> appMasterLocalResources = Maps.newHashMap();
+    // Add lib jars, and jars and files that the ApplicationMaster need as LocalResources
+    addAppMasterLocalResources(applicationId, appMasterLocalResources);
+
+    ContainerLaunchContext amContainerLaunchContext = Records.newRecord(ContainerLaunchContext.class);
+    amContainerLaunchContext.setLocalResources(appMasterLocalResources);
+    amContainerLaunchContext.setEnvironment(getEnvironmentVariables());
+    amContainerLaunchContext.setCommands(Lists.newArrayList(buildApplicationMasterCommand(resource.getMemory())));
+    if (UserGroupInformation.isSecurityEnabled()) {
+      setupSecurityTokens(amContainerLaunchContext);
+    }
+
+    // Setup the application submission context
+    appSubmissionContext.setApplicationName(this.appName);
+    appSubmissionContext.setResource(resource);
+    appSubmissionContext.setQueue(this.appQueueName);
+    appSubmissionContext.setPriority(Priority.newInstance(0));
+    appSubmissionContext.setAMContainerSpec(amContainerLaunchContext);
+
+    // Also setup container local resources by copying local jars and files the container need to HDFS
+    addContainerLocalResources(applicationId);
+
+    // Submit the application
+    LOGGER.info("Submitting application " + applicationId);
+    this.yarnClient.submitApplication(appSubmissionContext);
+
+    return applicationId;
+  }
+
+  private Resource prepareContainerResource(GetNewApplicationResponse newApplicationResponse) {
```

Comment:
Actually, I would think this method can be moved to a util class, since we would need similar code when specifying the resources for any new container.